packages/eslint-plugin/src/rules/no-unnecessary-condition.ts

```diff
@@ -294,6 +303,32 @@ export default createRule<Options, MessageId>({
       }
     }
 
+    function isCalleeAnAssertTypePredicate(
+      node: TSESTree.Node | undefined,
+    ): boolean {
+      if (!node || !isFunctionArgument(node)) {
+        return false;
+      }
+
+      const tsNode = service.esTreeNodeToTSNodeMap.get(node.callee);
+      const valueDeclaration = checker
+        .getTypeAtLocation(tsNode)
+        .getSymbol()?.valueDeclaration;
+
+      if (
+        !valueDeclaration ||
+        !ts.isFunctionDeclaration(valueDeclaration) ||
+        !valueDeclaration.type
+      ) {
+        return false;
+      }
+
+      return (
+        ts.isTypePredicateNode(valueDeclaration.type) &&
+        valueDeclaration.type.assertsModifier != null
```

Comment:
If I remove this second check the unit tests still pass. Is there a case where `assertsModifier` is ever not defined? 

* If there is, then let's add a test for it
* If not, let's remove this line (and maybe file an issue somewhere about the odd type?)

FWIW I can't figure out how to make it not-defined...

Reply:
> Is there a case where assertsModifier is ever not defined 

Honestly I don't remember at this point – I will look into this and write some detailed explanation, okay?